File: Question_8/Python/Question8.py
# Participants may update the following function parameters
def countNumberOfWays(numOfUnits, numOfCoinTypes, coins):
    table = [0] * (numOfUnits + 1)
    table[0] = 1
    for i in range(numOfCoinTypes):
        for j in range(coins[i], numOfUnits + 1):
            table[j] += table[j - coins[i]]
    return table[numOfUnits]


def main():
    firstLine = input().split(" ")
    secondLine = input().split(" ")

    numOfUnits = int(firstLine[0])
    numOfCoinTypes = int(firstLine[1])
    coins = list(map(int, secondLine))

    # Participants may update the following function parameters
    answer = countNumberOfWays(numOfUnits, numOfCoinTypes, coins)

    # Please do not remove the below line.
    print(answer)
    # Do not print anything after this line


if __name__ == '__main__':
    main()

File: Question_8/Java/Question8.java
import java.io.*;
import java.util.*;
import java.util.stream.Stream;

class Solution {
    static long countNumberOfWays(int numOfUnits, int numOfCoinTypes, int[] coins) {
        long[] table = new long[numOfUnits + 1];
        Arrays.fill(table, 0);
        table[0] = 1;
        for (int i = 0; i < numOfCoinTypes; i++)
            for (int j = coins[i]; j <= numOfUnits; j++)
                table[j] += table[j - coins[i]];
        return table[numOfUnits];
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        String[] firstLine = scanner.nextLine().split(" ");
        String[] secondLine = scanner.nextLine().split(" ");
        scanner.close();

        int numOfUnits = Integer.parseInt(firstLine[0]);
        int numOfCoinTypes = Integer.parseInt(firstLine[1]);
        int[] coins = Stream.of(secondLine).mapToInt(Integer::parseInt).toArray();

        long result = countNumberOfWays(numOfUnits, numOfCoinTypes, coins);
        // Please do not remove the below line.
        System.out.println(result);
        // Do not print anything after this line
    }
}
